architecture/2025-03-upgrade-design-doc.md

```diff
@@ -0,0 +1,793 @@
+| **Pre-flight Checks**    | Validation steps performed before an upgrade to ensure prerequisites are met and the system is in a valid state for upgrading.                                                                                                                                                                                                                                               |
+| **Version Skipping**     | The ability to upgrade directly from one version to a newer non-adjacent version (e.g., v0.40 → v0.44) without installing intermediate versions.                                                                                                                                                                                                                             |
+
+## Assumptions
+
+1. **User permissions**: Users running the upgrade command have enough permissions on both the Kubernetes cluster and the Radius installation.
+
+2. **Resource requirements**: The Kubernetes cluster has sufficient compute resources (CPU, memory) to run both the existing and new version components during the rolling upgrade process.
+
+3. **CLI version compatibility**: User has a Radius CLI version that includes the `rad upgrade kubernetes` feature. While older CLIs can't perform upgrades, newer CLIs maintain backward compatibility with older control planes.
+
+   Example: If you have CLI v0.42 and Control Plane v0.42:
+
+   - You CAN upgrade to Control Plane v0.44 using the v0.42 CLI (if v0.42 CLI includes the upgrade feature)
+   - After upgrading, your CLI (v0.42) will still work with basic operations against your v0.44 Control Plane
+   - However, you won't be able to use any new v0.44 features from your v0.42 CLI
+   - For full functionality, you should upgrade your CLI to match the Control Plane version after the upgrade completes
+
+4. **Network connectivity**: The upgrade process requires internet connectivity to pull container images from registries and validate versions. The process is not intended to run in air-gapped environments.
+
+   - An air-gapped environment is one where systems are physically isolated from unsecured networks like the public internet.
+   - These environments are common in high-security scenarios (military, financial, healthcare, government) where external network connectivity is restricted.
+
+5. **Stable starting state**: The Radius installation being upgraded is in a healthy, stable state. Attempting to upgrade an already failing installation may lead to unpredictable results.
+
+6. **Storage availability**: The Kubernetes cluster has sufficient persistent storage capacity for backup operations during the upgrade process.
+
+## Objectives
+
+> **Issue Reference:** <https://github.com/radius-project/radius/issues/8095>
+
+### Goals
+
+- **Simplify upgrade process**: Provide a single CLI command (`rad upgrade kubernetes`) to upgrade Radius without manual reinstallation.
```

Comment:
Will there be an option to upgrade via GitOps if the user manages their Radius installation via Flux?

Reply:
To be honest, I haven't explored this option. If GitOps is using Helm upgrade behind the scenes, it may be not that difficult to add. But I would discuss this with @willdavsmith and decide what to do next.